File: competitions/UCPC/2018/quals/D/D_tree_escape.py
# Wrong code, don't know why

from collections import deque

N = int(input())
edge = [[] for _ in range(N+1)]

for _ in range(N-1):
  a, b = (int(x) for x in input().split())
  edge[a].append(b)
  edge[b].append(a)

# DFS
stack = [(1, 0)] # (node, distance)
visited = [False for _ in range(N+1)]

win = False

while(len(stack) > 0):
  node, distance = stack.pop()
  cnt_unvisited = 0
  for adjacent in edge[node]:
    if not visited[adjacent]:
      stack.append((adjacent, distance + 1))
      visited[adjacent] = True
      cnt_unvisited += 1
  if cnt_unvisited == 0: # Leaf node
    ## print('({}, {})'.format(node, distance))
    win ^= (distance % 2 == 1)

if win:
  print('Yes')
else:
  print('No')

File: competitions/UCPC/2018/quals/D/D_tree_escape.cc
#include <stdio.h>
#include <vector>
using namespace std;

#define MAX_N 500000
#define TRUE 1
#define FALSE 0

int N;
vector<int> edge[MAX_N+1];
int node_stack[MAX_N+1];
int distance_stack[MAX_N+1];
int pointer = 0;
int visited[MAX_N+1];

int main() {
  scanf("%d", &N);

  for (int i = 1; i <= N - 1; i++) {
    int a, b;
    scanf("%d %d", &a, &b);
    edge[a].push_back(b);
    edge[b].push_back(a);
  }

  // DFS
  node_stack[0] = 1;
  distance_stack[0] = 0;
  visited[1] = TRUE;

  int win = FALSE;

  while (pointer >= 0) {
    int node = node_stack[pointer];
    int distance = distance_stack[pointer];
    pointer--;
    int cnt_unvisited = 0;

    for (int i = 0; i < edge[node].size(); i++) {
      int adjacent = edge[node][i];
      if (!visited[adjacent]) {
        pointer++;
        node_stack[pointer] = adjacent;
        distance_stack[pointer] = distance + 1;
        visited[adjacent] = TRUE;
        cnt_unvisited++;
      }
    }
    if (cnt_unvisited == 0) { // Leaf node
      // printf("(%d, %d)\n", node, distance);
      win ^= (distance % 2 == 1);
    }
  }

  if (win) {
    printf("Yes\n");
  } else {
    printf("No\n");
  }

  return 0;
}
